src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,2423 @@
+            BEAST_EXPECT(env.balance(bob) == preBob);
+        }
+    }
+
+    void
+    testOnlyOne(FeatureBitset features)
+    {
+        testcase("only one");
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        test::jtx::Env env{*this, envconfig()};
+
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+        auto const carol = Account("carol");
+        env.fund(XRP(1000), alice, bob, carol);
+        env.close();
+
+        auto const preAlice = env.balance(alice);
+        auto const preBob = env.balance(bob);
+
+        auto const batchFee = calcBatchFee(env, 0, 3);
+        auto const seq = env.seq(alice);
+        env(batch::outer(alice, seq, batchFee, tfOnlyOne),
+            batch::inner(pay(alice, bob, XRP(999)), seq + 1),
+            batch::inner(pay(alice, bob, XRP(1)), seq + 2),
+            batch::inner(pay(alice, bob, XRP(2)), seq + 3),
+            ter(tesSUCCESS));
+        auto const txIDs = env.tx()->getBatchTransactionIDs();
+        TxID const parentBatchId = env.tx()->getTransactionID();
+        std::vector<TestBatchData> testCases = {
+            {"tecUNFUNDED_PAYMENT", to_string(txIDs[0])},
+            {"tesSUCCESS", to_string(txIDs[1])},
+        };
+        env.close();
+        validateBatch(env, parentBatchId, testCases);
+
+        // Alice consumes sequences (# of txns)
+        BEAST_EXPECT(env.seq(alice) == seq + 3);
+
+        // Alice pays XRP & Fee; Bob receives XRP
+        BEAST_EXPECT(env.balance(alice) == preAlice - XRP(1) - batchFee);
+        BEAST_EXPECT(env.balance(bob) == preBob + XRP(1));
+    }
+
+    void
+    testUntilFailure(FeatureBitset features)
+    {
+        testcase("until failure");
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        test::jtx::Env env{*this, envconfig()};
+
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+        auto const carol = Account("carol");
+        env.fund(XRP(1000), alice, bob, carol);
+        env.close();
+
+        auto const preAlice = env.balance(alice);
+        auto const preBob = env.balance(bob);
+
+        auto const batchFee = calcBatchFee(env, 0, 4);
+        auto const seq = env.seq(alice);
+        env(batch::outer(alice, seq, batchFee, tfUntilFailure),
+            batch::inner(pay(alice, bob, XRP(1)), seq + 1),
+            batch::inner(pay(alice, bob, XRP(2)), seq + 2),
+            batch::inner(pay(alice, bob, XRP(999)), seq + 3),
+            batch::inner(pay(alice, bob, XRP(3)), seq + 4),
+            ter(tesSUCCESS));
+        auto const txIDs = env.tx()->getBatchTransactionIDs();
+        TxID const parentBatchId = env.tx()->getTransactionID();
+        std::vector<TestBatchData> testCases = {
+            {"tesSUCCESS", to_string(txIDs[0])},
+            {"tesSUCCESS", to_string(txIDs[1])},
+            {"tecUNFUNDED_PAYMENT", to_string(txIDs[2])},
```

Comment:
do we also have a test where the inner txn fails with with a tem code for tfUntilFailure?

Reply:
fixed: https://github.com/XRPLF/rippled/pull/5060/commits/6e38be1a1dda2bf83a3cf0e9b1788b75f460a808